sub-packages/bionemo-dnadl/src/bionemo/dnadl/io/dataset.py

```diff
@@ -0,0 +1,189 @@
+
+class GenomeDataset(Dataset, ABC):
+    """Abstract class for datasets that return a DNA sequence."""
+
+    def __init__(self, tokenizer: DNATokenizer):
+        """Instantiate the class."""
+        self.tokenizer = tokenizer
+
+    def __getitem__(self, ind: int) -> torch.tensor:
+        """Get an item from the dataset."""
+        sequence = self.get_dna_sequence(ind)
+        return self.tokenizer(sequence)
+
+    def get_dna_sequence(self, ind: int) -> str:
+        """Get the DNA sequence."""
+        ...
+
+
+class GenomeIntervalDataset(GenomeDataset):
+    """Datasets that extract sequences from a genome based on a bed file."""
+
+    def __init__(
+        self,
+        genome: Genome,
+        bed_file_df: pl.DataFrame,
+        context_length: int | None = None,
+        chr_bed_to_fasta_map: dict = {},
+        **kwargs: Any,
+    ):
+        """Instantiate the class.
+
+        Args:
+            genome: Genome object
+            bed_file_df: A bed file in the form of a polars DataFrame with chrom, start, and end columns at least
+            context_length: Size of the window
+            chr_bed_to_fasta_map: Mapping between chromosome names in the bed file and the fasta file
+            **kwargs: Additional arguments
+        """
+        super().__init__(**kwargs)
+        assert {"chrom", "start", "end"}.issubset(
+            bed_file_df.columns
+        ), "bed file df must contain columns ['chrom', 'start', 'end']"
+        self.bed_file_df = bed_file_df
+
+        # if the chromosome name in the bed file is different than the keyname in the fasta
+        # can remap on the fly
+        self.chr_bed_to_fasta_map = chr_bed_to_fasta_map
+        self.context_length = context_length
+        self.genome = genome
+
+    def __len__(self) -> int:
+        """Return the length of the dataset."""
+        return len(self.bed_file_df)
+
+    def get_interval(self, ind: int) -> GenomeInterval:
+        """Get the interval from the bed file."""
+        chrom = self.chr_bed_to_fasta_map.get(self.bed_file_df["chrom"][ind], self.bed_file_df["chrom"][ind])
+        genome_interval = GenomeInterval(chrom, self.bed_file_df["start"][ind], self.bed_file_df["end"][ind])
+        if self.context_length is not None:
+            return self.genome.adjust_interval(genome_interval, self.context_length)
+        return genome_interval
+
+    def get_dna_sequence(self, ind: int) -> str:
+        """Get the DNA sequence from the genome."""
+        genome_interval = self.get_interval(ind)
+        return self.genome.extract_sequence(genome_interval)
+
+
+class VCFDataset(GenomeDataset):
+    """Dataset that creates sequences based on a VCF. It needs a genome and a bed file to specify the windows."""
+
+    def __init__(
+        self,
+        vcf_file: str | Path,
+        genome_interval_dataset: GenomeIntervalDataset,
+        sample_ids: Optional[List[SampleId]] = None,
+        **kwargs: Any,
+    ):
+        """Initialize the class."""
```

Comment:
can we get docstrings for the arguments? Why is an interval dataset required? What does this do? Just guessing- its applying variants in the VCF to any position contained in the regions? Let's provide some examples in the class docstring.

Reply:
We might also want to make this class name less generic. It's a VCF dataset, but its doing something more specific- applying variants to fixed intervals.

There are models out there that are interested in windows centered around a variant (like `[NNNNNNNANNNNNN]`, which would have a difficult time benefitting from this class.

Reply:
Fair point, one can use site-only VCF for that. What do you prefer for the one with intervals GenotypedVCFDataset, or FixedIntervalsVCFDataset? I can also add a VCF dataset that looks only at the sites, each sample being centered on the variant